[Parser] Improving diagnostics for access-level modifiers

Visckmart opened the pull request:
## Problem
When declaring a method with multiple access-level modifiers (a programming mistake), the current diagnostic is not clear nor accurate.

As an example, when we compile the following block of code:
```Swift
class Foo {
  public private func exampleMethod() { }
}
```
We get the following diagnostics:
```Swift
error: duplicate modifier
  public private func exampleMethod() { }
         ^~~~~~~
note: modifier already specified here
  public private func exampleMethod() { }
  ^~~~~~
```

It can be argued that both modifiers are from the same "type" (access-level modifiers), but the example shows a situation where they aren't a duplicate of each other as the error message says.

## Solution
The solution implemented on this pull request is a new diagnostic: an error message that describes the problem more clearly and a note that refers to the previous modifier by name.
```Swift
error: multiple access-level modifiers specified
  public private func exampleMethod() { }
         ^~~~
note: 'public' previously specified here
  public private func exampleMethod() { }
  ^~~~~~~
```

Resolves #60914.

Visckmart committed:
[Parser] Improve diagnostics for access-level modifiers

Improve the diagnostics for situations where multiple access-level modifiers are used on the same declaration. Keep the original duplicate error message if the access levels are the same.

AnthonyLatsis commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -1485,6 +1485,10 @@ ERROR(duplicate_attribute,none,
+ERROR(multiple_access_level_modifiers,none,
+      "multiple access-level modifiers specified", ())
+NOTE(previous_access_level_modifier,none,
+     "'%0' previously specified here", (StringRef))
```
This is reminiscent of other "previously" notes like `multiple_override_prev` and reads as if `%0` was specified multiple times to me. I wish we could get away with just highlighting both modifiers, but our old diagnostics formatter can only highlight stuff on a single line. How about `previous modifier specified here`?

Visckmart replied:
Your suggestion looks good to me, but considering your previous suggestion of adding the word "incompatible" to the error message, what do you think of the following?
`incompatible modifier specified here`

AnthonyLatsis replied:
> what do you think of the following?
> `incompatible modifier specified here`

I incline to the neutral variant. "incompatible" here makes the situation sound like a one-way problem, like the previous attribute is "the bad guy".

Visckmart replied:
Got it!

AnthonyLatsis commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2926,9 +2926,25 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+                 PrevAccessControlAttr->getAttrName())
+          .highlight(PrevAccessControlAttr->getRange());
+        
+        // Removes the reference to the duplicate attribute as we don't need the
+        // extra diagnostic.
+        DuplicateAttribute = nullptr;
```
I think this is redundant because you deleted the line that was setting it for access-level modifiers.

Visckmart replied:
It's not redundant because this variable is set on the beginning of the function body. If I keep it set, it will trigger the original "Duplicate modifier" diagnostic.

AnthonyLatsis replied:
Oh, I see, `DuplicateAttribute` is supposed to be a constant. Could you adjust the logic to use the `DiscardAttribute` flag, as other cases do, and remove the mutations?

Visckmart replied:
At the beginning of the `case` there's one reassignment that looks unnecessary and I removed it now, as requested.

But I'm not sure `DuplicateAttribute` is supposed to be a constant. The original logic reassigns it to check for attributes like `private(set)`, that wouldn't be checked otherwise. So, I kept and improved it by writing a comment and updating the `DiscardAttribute` boolean, that was becoming outdated after the reassignment. Let me know what you think about it.

AnthonyLatsis commented:
@ahoppen Are you OK with accepting this sort of improvement into the legacy parser? I would guess `swift-syntax` is not really expected to diagnose this stuff anyway.

ahoppen commented:
> Are you OK with accepting this sort of improvement into the legacy parser? I would guess swift-syntax is not really expected to diagnose this stuff anyway.

Yes, swift-syntax allows duplicate access level modifiers because it just models them as a list. ASTGen will need to diagnose these sorts of problems. 

And I think this change is a good improvement. It would be nice to add a test case for it, though.

Visckmart commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -1485,6 +1485,10 @@ ERROR(duplicate_attribute,none,
 NOTE(previous_attribute,none,
      "%select{attribute|modifier}0 already specified here", (bool))
+ERROR(multiple_access_level_modifiers,none,
+      "multiple access-level modifiers specified", ())
```
Yeah! I even tested this case manually to make sure it worked as expected hahah Will make sure to change the diagnostic as you suggested.

Visckmart commented:
Cool! I updated the diagnostics' messages to reflect the latest from the discussion with @AnthonyLatsis and updated the existing tests to reflect the new diagnostics as @ahoppen requested.

AnthonyLatsis commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2986,9 +2986,26 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+      
+      if (!PrevAccessControlAttr) {
         Attributes.add(new (Context) AccessControlAttr(AtLoc, Loc, access));
+      } else {
```
The old duplication error is still preferable when they are actually duplicates, i.e. the parsed access levels match.

Visckmart replied:
Done!

AnthonyLatsis commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2986,9 +2986,26 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+        break;
+      }
+      
+      if (access != cast<AccessControlAttr>(*DuplicateAttribute).getAccess()) {
```
Does this compile? [`SetterAccessAttr::classof`](https://github.com/apple/swift/blob/4f68365467fcee6a25e39a07243b1668bb4388b8/include/swift/AST/Attr.h#L1171-L1173) suggests the cast source should be a pointer.

AnthonyLatsis replied:
I see now, LLVM has a template to allow for calling `cast` with a reference. If the source is passed around by pointer, we prefer the template that takes and produces a pointer: `cast<AccessControlAttr>(DuplicateAttribute)`.

Visckmart replied:
It did compile, but your suggestion is more consistent with the rest of the codebase as you've shown me in our discussions. I improved it on the latest push I made.

AnthonyLatsis commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -3063,9 +3063,26 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
+      }
+      
+      if (access != cast<AccessControlAttr>(DuplicateAttribute)->getAccess()) {
+        // Diagnose if there's already an access control attribute on
+        // this declaration with a different access level.
```
This comment summarizes the entire statement, so I think it deserves to be hoisted. Ditto the other case.

Now, some people seem to prefer commenting on statements within their code blocks. I’m not a fan of this style for several reasons:
* It adds to the cognitive load and speed of reading code for me, both because I have to figure out if a comment within a code block pertains to the entire statement or not, and because I’m used to reading the statement before deciding whether to descend into its code block(s).
* If you fold a code block in an IDE, the associated comment disappears.

Visckmart replied:
Done! I agree in this case it made a lot of sense to hoist it. And if it's the common practice in the codebase, it would be another great reason to do it.

On the other PR we're having a discussion, I believe there's a great example of outer comments talking about when and why the block of code is run, and inner comments talking about what the block of code does when it runs. To me, it was a great way of documenting the code.

AnthonyLatsis replied:
> On the other PR we're having a discussion, I believe there's a great example of outer comments talking about when and why the block of code is run, and inner comments talking about what the block of code does when it runs. To me, it was a great way of documenting the code.

This, right?
```c++
// Only allowed to synthesize a throwing super.init() call if the init being
// checked is also throwing.
if (ctor->hasThrows()) {
  // Diagnose on nonthrowing or rethrowing initializer
  if (!fromCtor->hasThrows() || fromCtor->hasPolymorphicEffect(EffectKind::Throws)) {
    fromCtor->diagnose(diag::implicit_throws_super_init);
    return true; // considered an error
  }
}
```

I agree the `hasThrows` behavioral quirk is worth the additional inner clarification. And, given this is a simple code block performing a single action, I really like that you’re at least alluding to the "what" besides talking about the "why" and "when" in the outer comment. On a larger scale, it becomes so much simpler to wrap your head around the logic behind the code with this context arrangement.

AnthonyLatsis reviewed (commented):
Awesome! Just one optional stylistic suggestion.

AnthonyLatsis commented:
@swift-ci please smoke test macOS

AnthonyLatsis commented:
@rintaro ping

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
The change looks good to me. But I feel these all attribute diagnostics including others (e.g. "concurrency only") should be in Sema (TypeCheckAttr.cpp) not in Parse or ASTGen. For example, `@differentiable` only diagnose duplication if the parameter is semantically the same. WDYT @ahoppen ?

It's not a blocker for merging this PR though.

ahoppen commented:
I agree with @rintaro. Implementing these checks in the type checker will make them applicable to both the old parser and the new parser.

ahoppen commented:
@swift-ci Please test Windows

Visckmart commented:
I just noticed I had mistakenly kept an extra argument on the note diagnostic. It's fixed now. I also ran the code formatter and cleaned up the commit message to make it more succinct.

Sorry for making these changes after the review comments and the tests have been run. @AnthonyLatsis would you mind triggering the tests once more?

AnthonyLatsis commented:
@swift-ci please smoke test

@rintaro Do we want anyone else to take a look at this?

AnthonyLatsis commented:
@swift-ci please smoke test

hborla reviewed (approved):
Thank you for your contribution! Exploring moving some of these attribute diagnostics out of the parser and into TypeCheckAttr.cpp would be a good next task if you're interested in continuing contributions in this area 🙂
